circle/src/cfft.rs

```diff
@@ -0,0 +1,218 @@
+use alloc::rc::Rc;
+use alloc::vec::Vec;
+use core::cell::RefCell;
+
+use p3_commit::PolynomialSpace;
+use p3_dft::divide_by_height;
+use p3_field::extension::{Complex, ComplexExtendable};
+use p3_field::{AbstractField, PackedValue};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::{Matrix, MatrixRowChunksMut, MatrixRowSlices, MatrixRowSlicesMut};
+use p3_maybe_rayon::prelude::*;
+use p3_util::log2_strict_usize;
+use tracing::instrument;
+
+use crate::domain::CircleDomain;
+use crate::twiddles::TwiddleCache;
+
+#[derive(Default, Clone)]
+pub struct Cfft<F>(Rc<RefCell<TwiddleCache<F>>>);
+
+impl<F: ComplexExtendable> Cfft<F> {
+    pub fn cfft(&self, vec: Vec<F>) -> Vec<F> {
+        self.cfft_batch(RowMajorMatrix::new_col(vec)).values
+    }
+    pub fn cfft_batch<M: MatrixRowChunksMut<F>>(&self, mat: M) -> M {
+        let log_n = log2_strict_usize(mat.height());
+        self.coset_cfft_batch(mat, F::circle_two_adic_generator(log_n + 1))
+    }
+    #[instrument(skip_all, fields(dims = %mat.dimensions()))]
+    pub fn coset_cfft_batch<M: MatrixRowChunksMut<F>>(&self, mut mat: M, shift: Complex<F>) -> M {
+        let n = mat.height();
+        let log_n = log2_strict_usize(n);
+
+        let mut cache = self.0.borrow_mut();
+        let twiddles = cache.get_twiddles(log_n, shift, true);
+
+        for (i, twiddle) in twiddles.iter().enumerate() {
+            let block_size = 1 << (log_n - i);
+            let half_block_size = block_size >> 1;
+            assert_eq!(twiddle.len(), half_block_size);
+
+            mat.par_row_chunks_mut(block_size).for_each(|mut chunk| {
+                for (i, &t) in twiddle.iter().enumerate() {
+                    let (lo, hi) = chunk.row_pair_slices_mut(i, block_size - i - 1);
+                    let (lo_packed, lo_suffix) = F::Packing::pack_slice_with_suffix_mut(lo);
+                    let (hi_packed, hi_suffix) = F::Packing::pack_slice_with_suffix_mut(hi);
+                    butterfly(lo_packed, hi_packed, t.into());
+                    butterfly(lo_suffix, hi_suffix, t);
+                }
+            });
+        }
+        divide_by_height(&mut mat);
+        mat
+    }
+
+    pub fn icfft(&self, vec: Vec<F>) -> Vec<F> {
+        self.icfft_batch(RowMajorMatrix::new_col(vec)).values
+    }
+    pub fn icfft_batch<M: MatrixRowChunksMut<F>>(&self, mat: M) -> M {
+        let log_n = log2_strict_usize(mat.height());
+        self.coset_icfft_batch(mat, F::circle_two_adic_generator(log_n + 1))
+    }
+    #[instrument(skip_all, fields(dims = %mat.dimensions()))]
+    pub fn coset_icfft_batch<M: MatrixRowChunksMut<F>>(&self, mat: M, shift: Complex<F>) -> M {
+        self.coset_icfft_batch_skipping_first_layers(mat, shift, 0)
+    }
+    #[instrument(skip_all, fields(dims = %mat.dimensions()))]
+    pub fn coset_icfft_batch_skipping_first_layers<M: MatrixRowChunksMut<F>>(
```

Comment:
`coset_icfft_batch_skipping_first_layers` shouldn't have been `pub`, thanks, fixed.

> Why isn't this version scaled ?

See the last few sentences of https://en.wikipedia.org/wiki/Discrete_Fourier_transform#Definition. Depending on what profession you're in, sometimes the IDFT is or isn't scaled. We choose to scale it so that

`x = cfft(icfft(x)) = icfft(cfft(x))`